src/main/java/rx/observables/StringObservable.java

```diff
@@ -114,28 +117,106 @@ public void call(S resource) {
      *            internal buffer size
      * @return the Observable containing read byte arrays from the input
      */
-    public static Observable<byte[]> from(final InputStream i, final int size) {
+    public static Observable<byte[]> from(final InputStream is, final int size) {
         return Observable.create(new OnSubscribe<byte[]>() {
             @Override
-            public void call(Subscriber<? super byte[]> o) {
+            public void call(Subscriber<? super byte[]> subscriber) {
+               subscriber.setProducer(new InputStreamProducer(is, subscriber, size));
+            }
+        });
+    }
+    
+    private static class InputStreamProducer implements Producer {
+
+        private final AtomicLong requested = new AtomicLong(0);
+        private final InputStream is;
+        private final Subscriber<? super byte[]> subscriber;
+        private int size;
+        
+        InputStreamProducer(InputStream is, Subscriber<? super byte[]> subscriber, int size) {
+            this.is = is;
+            this.subscriber = subscriber;
+            this.size = size;
+        }
+        
+        @Override
+        public void request(long n) {
+            try {
+                if (requested.get() == Long.MAX_VALUE)
```

Comment:
I think RxJava's contract on request(n) isn't as strong as reactive-streams, therefore, request() should be thread-safe and non-racy. Here, if a concurrent request(1) and request(Long.MAX_VALUE); arrives, the one can start both cases and end up having negative request amount.